src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorLanguageServiceClient.ts

```diff
@@ -52,7 +53,8 @@ export class RazorLanguageServiceClient {
     public async getSemanticTokenLegend(): Promise<vscode.SemanticTokensLegend | undefined> {
         await this.ensureStarted();
 
-        const response = await this.serverClient.sendRequest<vscode.SemanticTokensLegend>('_vs_/textDocument/semanticTokensLegend', /*request param*/null);
```

Comment:
I remember feeling like we needed to do this at the time, but I can't for the life of me remember why we couldn't just [get it off the capabilities](https://github.com/dotnet/razor-tooling/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/RazorSemanticTokensEndpoint.cs#L67). If we can manage to use the capabilities instead of having this endpoint that would be preferred, hopefully whatever blocker we had before have been resolved?

Reply:
Hmm, the Razor language server is OOP in VS Code, so I don't see how we can get the legend off the capabilities without sending a request, unless I'm missing something?

Reply:
The response to the "initialize" request which is the first request on server startup [is an "InitializeResult"](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeResult), which [contains `SemanticTokensOptions`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokensOptions), which has [the legend](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokensLegend). Since the initialize request is the first request for every LS we will have the Legend available by the time VSCode needs to handle a Semantic Tokens request's response, the question is basically just how to pipe it where it's needed on the VSCode side from there.

I really wish Past-Ryan had documented why we made the decision to use the out-of-spec call. I suspect that SemanticTokens was still experimental and the options didn't exist on the `InitializeParams` object yet.

Reply:
I think the problem is that we don't send the initialize request ourselves. The request is sent and received by the VSCode LSP client, and they don't expose to us the result of the initialize request (at least to my knowledge).

Basically, on the Razor VSCode side, we kick off a call to start the server with certain capabilities, but we have little access to what's going on between the VSCode client and language server.

Reply:
Talked offline. We're pretty sure the `client` object should have the `InitializeResult` available by the time `onStarted` is called, and we can use that to get the legend instead of this old endpoint. At that point we can delete the endpoint.